Lift some PSA test functions to a separate module

gilles-peskine-arm opened the pull request:
Move some functions and macros from specific PSA test suites to separate files under `tests/src` and `tests/include` so that they can be used from other test suites.

This pull requests addresses the following areas:
* `exercise_key` and friends.
* Storage cleanup.

The reason I'm doing this refactoring is to use the moved code in new test suites for [testing the storage format](https://github.com/ARMmbed/mbedtls/issues/3738). The following are out of scope:
* Any further reorganization.
* Fixing bugs or limitations in the moved code (apart from some very minor opportunistic fixing that I did while I noticed the problem).
* Using the newly moved code in existing test suites where it would be warranted.

bensze01 reviewed (changes requested):
Aside from the other changes requested, the commit messages of the following two commits suggest the opposite of what they actually do:

The commit message of [becacfc](https://github.com/ARMmbed/mbedtls/pull/4141/commits/becacfcb758a1b8f72a3c52642ce44bd8ab73c38) should be:
```diff
--- a/CommitMessage
+++ b/CommitMessage
@@ -4,7 +4,7 @@
 somewhat. The resulting behavior changes are:

 * Always check the exported length against PSA_EXPORT_KEY_OUTPUT_SIZE,
-  not just for unstructured key types.
+  not just for structured key types.
 * Always complain if a key type is not explicitly covered, not just
   for public keys.

```
And the commit message of [d14a4f3](https://github.com/ARMmbed/mbedtls/pull/4141/commits/d14a4f300b700e7d2736e48e73cb6376d7f631c7) should be:
```diff
--- a/CommitMessage
+++ b/CommitMessage
@@ -2,7 +2,7 @@

 ARRAY_LENGTH has a portable but unsafe implementation, and a
 non-portable implementation that causes a compile-time error if the
-macro is accidentally used on an array.
+macro is accidentally used on a pointer.

 The safety check was only implemented for __GCC__-defining compilers,
 but the part that triggered the compile-time error was always used. It
```

tom-daubney-arm reviewed (dismissed):
So apart from the copy paste and typos that have already been pointed out in tests/include/test/psa_exercise_key.h, I can't see anything wrong with this PR and so have approved it.

gilles-peskine-arm commented on tests/include/test/psa_exercise_key.h:
```diff
@@ -111,6 +119,26 @@
+ * \param alg               The algorithm to use.
+ * \param input1            The first input to pass.
+ * \param input1_length     The length of \p input1 in bytes.
+ * \param input1            The first input to pass.
+ * \param input1_length     The length of \p input1 in bytes.
```
I wondered why clang `-Wdocumentation` didn't catch this. [It's because we only pass `-Wdocumentation` when in `library`](https://github.com/ARMmbed/mbedtls/issues/4166).

gilles-peskine-arm commented on tests/src/psa_exercise_key.c:
```diff
@@ -707,20 +688,28 @@ int mbedtls_test_psa_exported_key_sanity_check(
     }
-    else
 
+#if defined(MBEDTLS_DES_C)
```
It is reachable. DES is considered an unstructured type. (In retrospect this may or may not have been the right thing: it helps about as often as it hurts.)

gilles-peskine-arm commented:
I made new commits to fix the documentation issues. Since it helped here, I added `-Wdocumentation` to the cmake build when using clang (it was previously only passed in `library`), and fixed the preexisting issues that this uncovered. Further improvements to preexisting documentation and further tweaks to build options are out of scope.

The current version is in https://github.com/gilles-peskine-arm/mbedtls/tree/psa-test-functions-lift-4. I'll push a new history that's identical except for fixing the mistakes in commit messages.

tom-daubney-arm reviewed (approved):
Thanks for the doc changes. LGTM.

bensze01 reviewed (dismissed):
Looks good to me. We just need a rebase now, due a conflict with upstream.

gilles-peskine-arm commented:
I rebased on top of `development` to resolve the add/add conflict in `tests/include/test/psa_crypto_helpers.h`. The previous version is at https://github.com/gilles-peskine-arm/mbedtls/tree/psa-test-functions-lift-5.

bensze01 reviewed (approved):
Looks good to me.

tom-daubney-arm reviewed (approved):
LGTM.
